cirq/circuits/circuit_dag.py

```diff
@@ -0,0 +1,157 @@
+    """A wrapper for a value that doesn't compare equal to other instances.
+
+    For example: 5 == 5 but Unique(5) != Unique(5).
+
+    Unique is used by CircuitDag to wrap operations because nodes in a graph
+    are considered the same node if they compare equal to each other.  X(q0)
+    in one moment of a Circuit and X(q0) in another moment of the Circuit are
+    wrapped by Unique(X(q0)) so they are distinct nodes in the graph.
+    """
+    def __init__(self, val: T) -> None:
+        self.val = val
+
+    def __repr__(self):
+        return 'Unique({}, {!r})'.format(id(self), self.val)
+
+
+TSelf = TypeVar('TSelf', bound='CircuitDag')
+
+
+def _default_can_reorder(op1: ops.Operation, op2: ops.Operation) -> bool:
+    """Returns true only if the operations have qubits in common."""
+    return not set(op1.qubits) & set(op2.qubits)
+
+
+class CircuitDag(networkx.DiGraph):
+    """A representation of a Circuit as a directed acyclic graph.
+
+    Nodes of the graph are instances of Unique containing each operation of a
+    circuit.
+
+    Edges of the graph are tuples of nodes.  Each edge specifies a required
+    application order between two operations.  The first must be applied before
+    the second.
+    """
+
+    default_can_reorder = staticmethod(_default_can_reorder)
+
+    def __init__(self,
+                 can_reorder: Callable[[ops.Operation, ops.Operation], bool] =
+                    _default_can_reorder,
+                 incoming_graph_data: Any = None,
+                 device: devices.Device = devices.UnconstrainedDevice
+                 ) -> None:
+        """Initializes a CircuitDag.
+
+        Args:
+            can_reorder: A predicate that determines if two operations may be
+                reordered.  Graph edges are created for pairs of operations
+                where this returns False.
+
+                The default predicate allows reordering only when the operations
+                don't share common qubits.
+            incoming_graph_data: Data in initialize the graph.  This can be any
+                value supported by networkx.DiGraph() e.g. an edge list or
+                another graph.
+            device: Hardware that the circuit should be able to run on.
+        """
+        super().__init__(incoming_graph_data)
+        self.can_reorder = can_reorder
+        self.device = device
+
+    @staticmethod
+    def make_node(op: ops.Operation) -> Unique:
+        return Unique(op)
+
+    @classmethod
+    def from_circuit(cls: Type[TSelf],
+                     circuit: circuit.Circuit,
+                     can_reorder: Callable[[ops.Operation, ops.Operation],
+                                           bool] = _default_can_reorder
+                     ) -> TSelf:
+        return cls.from_ops(circuit.all_operations(),
+                            can_reorder=can_reorder,
+                            device=circuit.device)
+
+    @classmethod
+    def from_ops(cls: Type[TSelf],
```

Comment:
Is it necessary for these to be class methods anymore, now that there is no inheritance present?

Reply:
All done.